engine/table/src/main/java/io/deephaven/engine/table/impl/UpdateBy.java

```diff
@@ -1,96 +1,842 @@
+                        final int size = update.shifted().size();
+
+                        // get these in ascending order and use a sequential builder
+                        for (int ii = 0; ii < size; ii++) {
+                            final long begin = update.shifted().getBeginRange(ii);
+                            final long end = update.shifted().getEndRange(ii);
+                            final long delta = update.shifted().getShiftDelta(ii);
+
+                            it.advance(begin);
+                            final RowSequence rs = it.getNextRowSequenceThrough(end);
+                            builder.appendRowSequenceWithOffset(rs, delta);
+                        }
+                        shiftedRows = builder.build();
+                    }
+                } else {
+                    shiftedRows = RowSetFactory.empty();
+                }
+            }
+
+            // this is where we leave single-threaded calls and rely on the scheduler to continue the work. Each
+            // call will chain to another until the sequence is complete
+            computeCachedColumnRowsets(
+                    () -> processWindows(
+                            () -> cleanUpAndNotify(
+                                    () -> {
+                                        // signal to the main task that we have completed our work
+                                        if (waitForResult != null) {
+                                            waitForResult.complete(null);
+                                        }
+                                    })));
+
+            if (waitForResult != null) {
+                try {
+                    // need to wait until this future is complete
+                    waitForResult.get();
+                } catch (InterruptedException e) {
+                    cleanUpAfterError();
+                    throw new CancellationException("interrupted while processing updateBy");
+                } catch (ExecutionException e) {
+                    cleanUpAfterError();
+                    if (e.getCause() instanceof RuntimeException) {
+                        throw (RuntimeException) e.getCause();
+                    } else {
+                        // rethrow the error
+                        throw new UncheckedDeephavenException("Failure while processing updateBy",
+                                e.getCause());
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * The Listener for apply to the constituent table updates
+     */
+    class UpdateByListener extends InstrumentedTableUpdateListenerAdapter {
+        public UpdateByListener(@Nullable String description) {
+            super(description, UpdateBy.this.source, false);
+        }
+
+        @Override
+        public void onUpdate(final TableUpdate upstream) {
+            final PhasedUpdateProcessor sm = new PhasedUpdateProcessor(upstream.acquire(), false);
```

Comment:
I need to research this, I originally added the `acquire()` call because it was prematurely being released but that may have been an artifact of a ListenerRecorder design pattern in use at the time.

Reply:
Need to `upstream.release()` in `cleanupAndNotify` and `cleanupAfterError`.

Reply:
You don't need the bucket results to notify ever. Delete these.
```

            // pass the update unchanged, just increment the ref count
            TableUpdate downstream = upstream.acquire();
            result.notifyListeners(downstream);
```